internal/pkg/policy/secret.go

```diff
@@ -39,57 +40,99 @@ func getSecretValues(ctx context.Context, secretRefs []model.SecretReferencesIte
 	if len(data.Inputs) == 0 {
-		return nil, nil
+		return nil, nil, nil
 	}
 
 	if len(data.SecretReferences) == 0 {
-		return data.Inputs, nil
+		return data.Inputs, nil, nil
 	}
 
 	secretValues, err := getSecretValues(ctx, data.SecretReferences, bulker)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	result := make([]map[string]interface{}, 0)
-	for _, input := range data.Inputs {
-		newInput := make(map[string]interface{})
-		for k, v := range input {
-			newInput[k] = replaceAnyRef(v, secretValues)
+	keys := make([]string, 0)
+	for i, input := range data.Inputs {
+		newInput, ks := replaceMapRef(input, secretValues)
+		for _, key := range ks {
+			keys = append(keys, fmt.Sprintf("inputs[%d].%s", i, key))
 		}
 		result = append(result, newInput)
 	}
 	data.SecretReferences = nil
-	return result, nil
+	return result, keys, nil
 }
 
-// replaceAnyRef is a generic approach to replacing any secret references in the passed item.
-// It will go through any slices or maps and replace any secret references.
-//
-// go's generic parameters are not a good fit for rewriting this method as the typeswitch will not work.
-func replaceAnyRef(ref any, secrets map[string]string) any {
+// replaceMapRef replaces all nested secret values in the passed input and returns the resulting input along with a list of keys where inputs have been replaced.
+func replaceMapRef(input map[string]any, secrets map[string]string) (map[string]any, []string) {
+	keys := make([]string, 0)
+	result := make(map[string]any, len(input))
 	var r any
-	switch val := ref.(type) {
-	case string:
-		r = replaceStringRef(val, secrets)
-	case map[string]any:
-		obj := make(map[string]any)
-		for k, v := range val {
-			obj[k] = replaceAnyRef(v, secrets)
+
+	for k, v := range input {
+		switch value := v.(type) {
+		case string:
+			ref, replaced := replaceStringRef(value, secrets)
+			if replaced {
+				keys = append(keys, k)
+			}
+			r = ref
+		case map[string]any:
+			ref, ks := replaceMapRef(value, secrets)
+			for _, key := range ks {
+				keys = append(keys, k+"."+key)
+			}
+			r = ref
+		case []any:
+			ref, ks := replaceSliceRef(value, secrets)
+			for _, key := range ks {
+				keys = append(keys, k+key)
+			}
+			r = ref
+		default:
+			r = v
 		}
-		r = obj
-	case []any:
-		arr := make([]any, len(val))
-		for i, v := range val {
-			arr[i] = replaceAnyRef(v, secrets)
+		result[k] = r
+	}
+	return result, keys
+}
+
+// replaceSliceRef replaces all nested secrets within the passed slice and returns the resulting slice along with a list of keys that indicate where values have been replaced.
+func replaceSliceRef(arr []any, secrets map[string]string) ([]any, []string) {
+	keys := make([]string, 0)
+	result := make([]any, len(arr))
+	var r any
+
+	for i, v := range arr {
+		switch value := v.(type) {
+		case string:
+			ref, replaced := replaceStringRef(value, secrets)
+			if replaced {
+				keys = append(keys, fmt.Sprintf("[%d]", i))
```

Comment:
I'm going to delay merging this as I think go-ucfg uses `.` when accessing array values.

Reply:
https://github.com/elastic/elastic-agent/pull/5621
go-ucfg does use `.` when accessing arrays